userguide/03. KeyedArtifacts and Joining.ipynb

```diff
@@ -0,0 +1,429 @@
+    "\n",
+    "For now, here is a quick demonstration of finding the key by attribute and by index:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "key = 0 ; [0] = 0 ; [1] = KeyedArtifact\n",
+      "key = 1 ; [0] = 1 ; [1] = KeyedArtifact\n",
+      "key = 2 ; [0] = 2 ; [1] = KeyedArtifact\n",
+      "----\n",
+      "0\n",
+      "100\n",
+      "200\n"
+     ]
+    }
+   ],
+   "source": [
+    "def lame_deriver(num):\n",
+    "    return num * 100\n",
+    "\n",
+    "chan = IterChannel(KeyedArtifact(i, DerivedArtifact(lame_deriver, i)) for i in range(3))\n",
+    "tee = chan.tee()\n",
+    "app.Flo([chan.map(lambda a: print('key =', a.key, '; [0] =', a[0], '; [1] =', a[1]))]).run()\n",
+    "print('----')\n",
+    "print_chans(tee)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Joining"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "One thing `KeyedArtifacts` make possible is joining between two channels using the keys of the items in each channel.\n",
+    "\n",
+    "flowz provides a `CoGroupChannel` that joins two channels, grouping the items by their keys in a very \"cover all bases\" approach that is reminiscent of a \"full outer join\" in SQL.  Its default approach can be difficult to understand (though it is documented well in the docstring for `CoGroupChannel`), but it can be easily specialized.\n",
+    "\n",
+    "A very useful specialization is supplied by the `flowz.util.channel_inner_join(a, b)` function, which will join together the items in each channel that have equivalent keys.\n",
+    "\n",
+    "> NOTE: It is important that the channels to be joined already present their objects in the relevant \"sort\" order for their keys.  Sort operations on channels are conspicuously missing from flowz -- on purpose -- because they go against the philosophy of low-overhead streaming of data via channel.  Sorting would require creating all the artifacts in a channel, holding them in memory, sorting them by key, then presenting them to the consumer in sorted order.  Everything else in flowz creates new artifacts only at the time they are needed, and they have potentially short lifetimes."
```

Comment:
It is true that sorting operations are missing, and that's by design.  But the opinions behind that aren't quite as strong as you might suppose.

I _expect_ that we'll add sorting at some point.  It's a fairly obvious gap in the feature set, and it's fairly easily expressed directly with existing functions:

```python
sorted_chan = unsorted_chan.groupby(lambda _: 1)[1].flat_map(sorted)
```

I didn't add sorting primarily because I knew for initial uses that all such sorting could be expressed at origin points (at points where I'm selecting data from various sources), and I figuring on punting anything that wasn't strictly necessary.  I also figuring I would eventually build windowing/grouping, at which point adding sorting support would be simple.

Your arguments about sorting and memory usage and whatnot are all technically true, but it wasn't my intention for flowz to be like "no, you can't do that."